image/dicom_le.ksy

```diff
@@ -0,0 +1,147 @@
+meta:
+  id: dicom
+  file-extension: dcm
+  endian: le
+seq:
+  - id: file_header
+    type: t_file_header
+  - id: dataset
+    type: t_dataset
+types:
+  t_file_header:
+    seq:
+      - id: preamble
+        size: 128
+      - id: magic
+        contents: 'DICM'
+  t_dataset:
+    seq:
+      - id: elements1
+        type: t_dataentry_state_implicit
+        if: not _io.eof
+        repeat: until
+        repeat-until: _io.eof or _.is_transfer_syntax_change_explicit
+      - id: elements2
+        type: t_dataentry_state_explicit
+        repeat: eos
+  t_dataentry_state_implicit:
+    seq:
+      - id: tag_group
+        type: u2
+      - id: tag_elem
+        type: u2
+      - id: header
+        type: t_entry_header_implicit
+        if: entry_implicit
+      - id: header
+        type: t_entry_header_explicit
+        if: not entry_implicit
+      - id: content
+        size: header.content_length
+        if: has_content
+      - id: elements
+        type: t_dataentry_state_implicit
+        if: has_elements and not is_definite
+        repeat: until
+        repeat-until: _.is_closing_tag
+      - id: elements
+        type: t_elements
+        if: has_elements and is_definite
+        size: header.content_length
+    types:
+      t_elements:
+        seq:
+          - id: elements
+            type: t_dataentry_state_implicit
+            repeat: eos
+    instances:
+      entry_implicit:
+        value: tag_group != 2 or tag_group == 0xfffe
+      is_closing_tag:
+        value: tag_group == 0xfffe and tag_elem & 0xff00 == 0xe000 and tag_elem != 0xe000
+      has_elements:
+        value: (tag_group == 0xfffe and tag_elem == 0xe000) or header.is_seq
```

Comment:
Using `header_is.seq` is a bad idea even in dynamic languages, as there is no guarantee that `is_seq` attribute will exist. If it won't, it will result in runtime error. With our switching type, it's better to use check and a cast:

```yaml
      has_elements:
        value: (tag_group == 0xfffe and tag_elem == 0xe000) or (not entry_implicit and (header.as<t_entry_header_explicit>.is_seq))
```